ADBDEV-3333: Flow isn't set properly for Shared Scan nodes during translation to physical plan

RekGRpth opened the pull request:
`ORCA` does not always correctly fill in the `flow` for `shared scans`, although the `shareinput_mutator_xslice_*` functions in `cdbmutate` require a correctly filled `flow`.
These functions, among other things, check that [if at least one part of a `shared scan` is executed on the `coordinator`](https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/cdb/cdbmutate.c#L2560-L2565), then [all its other parts must also be performed on the `coordinator`](https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/cdb/cdbmutate.c#L2602-L2609).
They do this on the `shared scan` `flow`, assuming that if `flotype = FLOW_SINGLETON` and `segindex < 0`, then [the `shared scan` will be performed on the `coordinator`](https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/cdb/cdbmutate.c#L2419-L2423).
For such a check, [`flow` must be filled in for a `shared scan`](https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/cdb/cdbmutate.c#L2418).
[`ORCA` always fills the `flow` in the `shared scan`](https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp#L3725-L3729), but sometimes [it does this incorrectly](https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp#L3766-L3770).
[`ORCA` correctly fills the `flow` only under the `Motion` nodes](https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp#L2273), and if there is no `Motion` above the `producers` or `consumers` of the `shared scan`, then the `flow` of this `shared scan` will be filled incorrectly with the parameters `flotype = FLOW_UNDEFINED` and `segindex = 0`, which is not true if the `slice` will be executed only on one of the `segments` or even on the `coordinator`.
The `executor` determines [where the `slice` will be executed using `gangType` in the `FillSliceGangInfo` function](https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/executor/execMain.c#L4996-L5031), and [`flotype` and `segindex` are used to determine `gangType` only under `Motion` nodes](https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/executor/execMain.c#L5150-L5177).
This shows that the `flow` of a `shared scan` has no effect on where the `slice` will be executed, as for this, the `flow` of the node under the `Motion` of the `slice` in which the `shared scan` is located is used.
To determine where the `slice` will be executed, we can use the node's `flow` under `Motion`, keeping and checking that the `slice` is being executed on the `coordinator` by the conditions `flotype = FLOW_SINGLETON` and `segindex < 0`.
First we need to define where the root `slice` will be executed, which can be done in the same way as it is done in the `executor`.
The following cases are possible.
1) For `CTAS`, `SELECT INTO`, `COPY INTO`, `REFRESH MATERIALIZED VIEW`, [the root `slice` is not executed on the `coordinator`](https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/executor/execMain.c#L5231-L5247).
2) For the modification node, the table distribution policy is checked, and if the table is not master-only, then [the root `slice` is not executed on the `coordinator`](https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/executor/execMain.c#L5051-L5116).
3) Parallel cursors can only be generated by the postgres planner, which means [it will set the correct `flow`](https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/executor/execMain.c#L1809-L1843).
If it is not a `singleton`, then the root `slice` is not executed on the `coordinator`.
If this is a `singleton`, but the `locus` is not `Entry`, not `General`, and not `SingleQE`, then the root `slice` is also not executed on the `coordinator`.
4) The postgres planner fills `flow` correctly.
If it is not a `singleton`, then the root `slice` is not executed on the `coordinator`.
If this is a `singleton` on a segment, then the root `slice` is also not executed on the `coordinator`.
5) In all other cases, the root `slice` is executed on the `coordinator`.

We get rid of the need to fill in the `flow` for a `shared scan` in `ORCA`, and now we can remove the function that did this.

For example consider following query
```sql
CREATE TABLE d (a int, b int, c int) DISTRIBUTED BY (a);
CREATE TABLE r (a int, b int, c char(255)) DISTRIBUTED REPLICATED;
INSERT INTO d SELECT 1, generate_series(1,10), 1;
INSERT INTO r SELECT 1, 2, generate_series(1,100);
EXPLAIN (ANALYZE off, COSTS off, VERBOSE off)
WITH e AS (
    SELECT DISTINCT b FROM d
), h AS (
    SELECT a FROM d JOIN e f USING (b) JOIN e USING (b)
) SELECT * FROM r JOIN h USING (a) JOIN h i USING (a);
FATAL:  Unexpected internal error (cdbmutate.c:2606)
DETAIL:  FailedAssertion("!(plan->flow->flotype == FLOW_SINGLETON)", File: "cdbmutate.c", Line: 2606)
HINT:  Process 17639 will wait for gp_debug_linger=10 seconds before termination.
Note that its locks and other resources will not be released until then.
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
connection to server was lost
```
without assert it gives
```sql
EXPLAIN (ANALYZE off, COSTS off, VERBOSE off)
WITH e AS (
    SELECT DISTINCT b FROM d
), h AS (
    SELECT a FROM d JOIN e f USING (b) JOIN e USING (b)
) SELECT * FROM r JOIN h USING (a) JOIN h i USING (a);
                                              QUERY PLAN                                              
------------------------------------------------------------------------------------------------------
 Sequence
   ->  Shared Scan (share slice:id 0:0)
         ->  Materialize
               ->  Gather Motion 3:1  (slice7; segments: 3)
                     ->  GroupAggregate
                           Group Key: d_1.b
                           ->  Sort
                                 Sort Key: d_1.b
                                 ->  Redistribute Motion 3:3  (slice6; segments: 3)
                                       Hash Key: d_1.b
                                       ->  Seq Scan on d d_1
   ->  Sequence
         ->  Shared Scan (share slice:id 0:1)
               ->  Materialize
                     ->  Gather Motion 3:1  (slice5; segments: 3)
                           ->  Hash Join
                                 Hash Cond: (share0_ref3.b = d.b)
                                 ->  Redistribute Motion 1:3  (slice2)
                                       Hash Key: share0_ref3.b
                                       ->  Shared Scan (share slice:id 2:0)
                                 ->  Hash
                                       ->  Hash Join
                                             Hash Cond: (share0_ref2.b = d.b)
                                             ->  Redistribute Motion 1:3  (slice3)
                                                   Hash Key: share0_ref2.b
                                                   ->  Shared Scan (share slice:id 3:0)
                                             ->  Hash
                                                   ->  Redistribute Motion 3:3  (slice4; segments: 3)
                                                         Hash Key: d.b
                                                         ->  Seq Scan on d
         ->  Hash Join
               Hash Cond: (r.a = share1_ref3.a)
               ->  Gather Motion 1:1  (slice1; segments: 1)
                     ->  Seq Scan on r
               ->  Hash
                     ->  Hash Join
                           Hash Cond: (share1_ref3.a = share1_ref2.a)
                           ->  Shared Scan (share slice:id 0:1)
                           ->  Hash
                                 ->  Shared Scan (share slice:id 0:1)
 Optimizer: Pivotal Optimizer (GPORCA)
(41 rows)
EXPLAIN (ANALYZE on, COSTS off, VERBOSE off)
WITH e AS (
    SELECT DISTINCT b FROM d
), h AS (
    SELECT a FROM d JOIN e f USING (b) JOIN e USING (b)
) SELECT * FROM r JOIN h USING (a) JOIN h i USING (a);
                                                                   QUERY PLAN                                                                   
------------------------------------------------------------------------------------------------------------------------------------------------
 Sequence (actual time=9.494..68.605 rows=10000 loops=1)
   ->  Shared Scan (share slice:id 0:0) (actual time=1.873..1.888 rows=10 loops=1)
         ->  Materialize (actual time=1.851..1.851 rows=0 loops=1)
               ->  Gather Motion 3:1  (slice7; segments: 3) (actual time=1.437..1.766 rows=10 loops=1)
                     ->  GroupAggregate (actual time=1.596..1.612 rows=5 loops=1)
                           Group Key: d_1.b
                           ->  Sort (actual time=1.587..1.592 rows=5 loops=1)
                                 Sort Key: d_1.b
                                 Sort Method:  quicksort  Memory: 99kB
                                 ->  Redistribute Motion 3:3  (slice6; segments: 3) (actual time=1.206..1.220 rows=5 loops=1)
                                       Hash Key: d_1.b
                                       ->  Seq Scan on d d_1 (actual time=0.007..0.018 rows=10 loops=1)
   ->  Sequence (actual time=7.590..44.431 rows=10000 loops=1)
         ->  Shared Scan (share slice:id 0:1) (actual time=6.654..6.664 rows=10 loops=1)
               ->  Materialize (actual time=6.648..6.651 rows=0 loops=1)
                     ->  Gather Motion 3:1  (slice5; segments: 3) (actual time=6.316..6.629 rows=10 loops=1)
                           ->  Hash Join (actual time=8.006..8.582 rows=5 loops=1)
                                 Hash Cond: (share0_ref3.b = d.b)
                                 Extra Text: (seg0)   Hash chain length 1.0 avg, 1 max, using 5 of 131072 buckets.
                                 ->  Redistribute Motion 1:3  (slice2) (actual time=2.574..2.581 rows=5 loops=1)
                                       Hash Key: share0_ref3.b
                                       ->  Shared Scan (share slice:id 2:0) (actual time=0.039..0.054 rows=10 loops=1)
                                 ->  Hash (actual time=3.921..3.921 rows=5 loops=1)
                                       ->  Hash Join (actual time=3.493..3.908 rows=5 loops=1)
                                             Hash Cond: (share0_ref2.b = d.b)
                                             Extra Text: (seg0)   Hash chain length 1.0 avg, 1 max, using 5 of 131072 buckets.
                                             ->  Redistribute Motion 1:3  (slice3) (actual time=2.328..2.333 rows=5 loops=1)
                                                   Hash Key: share0_ref2.b
                                                   ->  Shared Scan (share slice:id 3:0) (actual time=0.039..0.052 rows=10 loops=1)
                                             ->  Hash (actual time=0.902..0.902 rows=5 loops=1)
                                                   ->  Redistribute Motion 3:3  (slice4; segments: 3) (actual time=0.008..0.889 rows=5 loops=1)
                                                         Hash Key: d.b
                                                         ->  Seq Scan on d (actual time=0.019..0.033 rows=10 loops=1)
         ->  Hash Join (actual time=0.908..15.473 rows=10000 loops=1)
               Hash Cond: (r.a = share1_ref3.a)
               Extra Text: Hash chain length 100.0 avg, 100 max, using 1 of 131072 buckets.
               ->  Gather Motion 1:1  (slice1; segments: 1) (actual time=0.018..0.237 rows=100 loops=1)
                     ->  Seq Scan on r (actual time=0.024..0.140 rows=100 loops=1)
               ->  Hash (actual time=0.798..0.798 rows=100 loops=1)
                     Buckets: 131072  Batches: 1  Memory Usage: 3kB
                     ->  Hash Join (actual time=0.138..0.651 rows=100 loops=1)
                           Hash Cond: (share1_ref3.a = share1_ref2.a)
                           Extra Text: Hash chain length 10.0 avg, 10 max, using 1 of 131072 buckets.
                           ->  Shared Scan (share slice:id 0:1) (actual time=0.004..0.016 rows=10 loops=1)
                           ->  Hash (actual time=0.036..0.036 rows=10 loops=1)
                                 Buckets: 131072  Batches: 1  Memory Usage: 1kB
                                 ->  Shared Scan (share slice:id 0:1) (actual time=0.007..0.020 rows=10 loops=1)
 Planning time: 742.673 ms
   (slice0)    Executor memory: 2594K bytes.  Work_mem: 3K bytes max.
   (slice1)    Executor memory: 62K bytes (seg2).
   (slice2)    Executor memory: 158K bytes (entry db).
   (slice3)    Executor memory: 158K bytes (entry db).
   (slice4)    Executor memory: 62K bytes avg x 3 workers, 62K bytes max (seg0).
   (slice5)    Executor memory: 2146K bytes avg x 3 workers, 2146K bytes max (seg0).  Work_mem: 1K bytes max.
   (slice6)    Executor memory: 62K bytes avg x 3 workers, 62K bytes max (seg0).
   (slice7)    Executor memory: 94K bytes avg x 3 workers, 94K bytes max (seg0).  Work_mem: 65K bytes max.
 Memory used:  256000kB
 Optimizer: Pivotal Optimizer (GPORCA)
 Execution time: 101.957 ms
(59 rows)
```
This query has two CTE's with producers `Shared Scan (share slice:id 0:0)` and `Shared Scan (share slice:id 0:1)`.
Both producers are executed on coordinator and part of consumers `Shared Scan (share slice:id 0:1)` and `Shared Scan (share slice:id 0:1)` are also executed on coordinator.
While other part of consumers `Shared Scan (share slice:id 2:0)` and `Shared Scan (share slice:id 3:0)` are executed on `entry db` in different slices `slice2` and `slice3`.
In this plan ORCA correctly fills flows for shared scans executed on `entry db` because they are placed under motions https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp#L2273
but for other shared scans (which have not motions above) ORCA sets incorrectly filled flow https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp#L3766-L3770
Then `shareinput_mutator_xslice_1` function builds global list `qdShares` of shared scan ids executing on coordinator https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/cdb/cdbmutate.c#L2418-L2423
In our case it is only `[0]`, because only share input scan with `id = 0` has correctly filled flow with `flotype = FLOW_SINGLETON` and `segindex = -1`, while share input scan with `id = 1` has incorrectly filled flow `flotype = FLOW_UNDEFINED` and `segindex = 0`.
Then `shareinput_mutator_xslice_3` function builds global list `qdSlices` of slice ids executing on coordinator for `qdShares` https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/cdb/cdbmutate.c#L2560-L2565
In our case it is `[0, 3, 2]`, because share input scan with `id = 0` (executing on coordinator) has slices 0, 2 and 3.
And finally, `shareinput_mutator_xslice_4` function requires `flotype` of all flows in `qdSlices` to be `FLOW_SINGLETON` always https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/cdb/cdbmutate.c#L2602-L2609
In our case it is not so for flow of share input scan `id = 1` where `flotype = FLOW_UNDEFINED` and therefore assert rises.
Greenplum detects where to execute `slice` by it `gangType` and respectively sets `gangSize` and `segments` in `FillSliceGangInfo` function https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/executor/execMain.c#L4996-L5031
While `flow` `flotype` and `segindex` are used to setup `gangType` only under `Motion` node in https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/executor/execMain.c#L5150-L5177
Therefore we need to ensure that `flow` for node under `Motion` is correctly filled and on coordinator it has `flotype = FLOW_SINGLETON` and `segindex = -1`.
It may be done without filling (incorrectly by ORCA) `flow` for shared scan, but instead we can save and check `segindex` of `flow` under `Motion` node.

Stolb27 commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -3763,12 +3763,6 @@ CTranslatorDXLToPlStmt::GetFlowCTEConsumer(List *shared_scan_cte_consumer_list)
-	{
-		flow = MakeNode(Flow);
-		flow->flotype = FLOW_UNDEFINED;	 // default flow
-	}
```
After these changes, this function may return null values that are used as flow for descendant plan nodes (Sort or Material). Why assignment of nulls is correct?

RekGRpth replied:
> this function may return null values that are used as flow for descendant plan nodes (Sort or Material). Why assignment of nulls is correct?

because only flow under Motion node is significant, not under shared scan

InnerLife0 replied:
I have another question about this block of code. Will it be naive to keep it and remove only one FLOW_SINGLETON assertion? By this, we can make patch less invasive.

RekGRpth replied:
> remove only one FLOW_SINGLETON assertion

was first idea

InnerLife0 replied:
So what is the reason of choosing a current option?

RekGRpth replied:
> So what is the reason of choosing a current option?

firstly, to not incorrectly fill flow by orca
and secondly, to to ensure that flow for node under Motion is correctly filled and on coordinator it has flotype = FLOW_SINGLETON and segindex = -1

RekGRpth replied:
I removed this function, but saved assert.

HustonMmmavr replied:
To summarize: the `flow` of `ShareInputScan/Material/Sort` nodes are not significant to make decision at which segments (segment(s) or coordinator) the slice will be executed.
So now `flow` of `Material` and `Sort` nodes are not filled and for `ShareInputScan` nodes there is an assertion which checks that `ShareInputScan` nodes (where `flow` is not NULL) must have the same `flotype`. 
I'm right?

RekGRpth replied:
> To summarize: the `flow` of `ShareInputScan/Material/Sort` nodes are not significant to make decision at which segments (segment(s) or coordinator) the slice will be executed. So now `flow` of `Material` and `Sort` nodes are not filled and for `ShareInputScan` nodes there is an assertion which checks that `ShareInputScan` nodes (where `flow` is not NULL) must have the same `flotype`. I'm right?

Yes

InnerLife0 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -2406,20 +2426,24 @@ shareinput_mutator_xslice_1(Node *node, PlannerInfo *root, bool fPop)
 		shareinput_pushmot(ctxt, motion->motionID);
+
+		/* Top node of subplan should have a Flow node. */
+		Insist(motion->plan.lefttree && motion->plan.lefttree->flow);
```
One of my PRs to upstream wasn't approved, because it contained Insist macro. The guys from upstream want to avoid using it for a reason, so, we should avoid its usage too.

RekGRpth replied:
> One of my PRs to upstream wasn't approved, because it contained Insist macro. The guys from upstream want to avoid using it for a reason, so, we should avoid its usage too.

I copy it from here https://github.com/arenadata/gpdb/blob/e69509affc7d96b7298b9ae80582c974cd4b0d4a/src/backend/executor/execMain.c#L5126-L5127

InnerLife0 replied:
So what? There are multiple examples of using Insist in code.

RekGRpth replied:
may be use `AssertImply`?

InnerLife0 replied:
I don't think it fit your needs. But maybe you can show the way to implement the same condition using AssertImply correctly.

RekGRpth replied:
> you can show the way to implement the same condition using AssertImply correctly

```c
AssertImply(motion->plan.lefttree, motion->plan.lefttree->flow); 
```
?

InnerLife0 replied:
Will assertion fire if motion->plan.lefttree is NULL?

RekGRpth replied:
or simply
```c
Assert(motion->plan.lefttree);
Assert(motion->plan.lefttree->flow); 
```

InnerLife0 replied:
Yes, that's one of options.

RekGRpth replied:
done

InnerLife0 commented:
The solution looks elegant in general.
The PR comment contains grammar mistakes which I don't want to fix this time. Take a look at [this](https://grammarist.com/usage/therefore-vs-therefor/) for example. Try to re-read some of your sentences. They just want to be rearranged.
I see you usually don't like to write full-fledged commit messages. Are you waiting for somebody to fix your PR description so you can write commit message based on it?

RekGRpth commented:
> Take a look at [this](https://grammarist.com/usage/therefore-vs-therefor/) for example

fixed

RekGRpth commented:
```
Sometimes ORCA does not fill flow for shared scan correctly,
while cdbmuate shareinput mutator xslice functions always
require to correctly filled flow.
ORCA correctly fills flows for shared scans executed on entry db
because they are placed under motions.
But for other shared scans (which have not motions above)
ORCA sets incorrectly filled flow.
We need to ensure that flow for node under Motion is correctly filled
and on coordinator it has flotype = FLOW_SINGLETON and segindex = -1.
It may be done without filling (incorrectly by ORCA) flow for shared scan,
but instead we can save and check segindex of flow under Motion node.
```

RekGRpth commented:
> commit messages

```
Sometimes ORCA does not fill flow for shared scan correctly,
while cdbmuate shareinput mutator xslice functions always
require to correctly filled flow.
ORCA correctly fills flows for shared scans executed on entry db
because they are placed under motions.
But for other shared scans (which have not motions above)
ORCA sets incorrectly filled flow.
We need to ensure that flow for node under Motion is correctly filled
and on coordinator it has flotype = FLOW_SINGLETON and segindex = -1.
It may be done without filling (incorrectly by ORCA) flow for shared scan,
but instead we can save and check segindex of flow under Motion node.
```

RekGRpth committed:
ORCA does not always correctly fill in the flow for shared scans,
although the shareinput_mutator_xslice_* functions in cdbmutate
require a correctly filled flow.
These functions, among other things, check that if at least one part
of a shared scan is executed on the coordinator, then all its other parts
must also be performed on the coordinator.
They do this on the shared scan flow, assuming that if
flotype = FLOW_SINGLETON and segindex < 0, then
the shared scan will be performed on the coordinator.
For such a check, flow must be filled in for a shared scan.
ORCA always fills the flow in the shared scan,
but sometimes it does this incorrectly.
ORCA correctly fills the flow only under the Motion nodes, and if
there is no Motion above the producers or consumers of the shared scan,
then the flow of this shared scan will be filled incorrectly
with the parameters flotype = FLOW_UNDEFINED and segindex = 0,
which is not true if the slice will be executed only on
one of the segments or even on the coordinator.
The executor determines where the slice will be executed using gangType in
the FillSliceGangInfo function, and flotype and segindex are used
to determine gangType only under Motion nodes.
This shows that the flow of a shared scan has no effect on
where the slice will be executed, as for this, the flow of the node
under the Motion of the slice in which the shared scan is located is used.
To determine where the slice will be executed, we can use the node's flow
under Motion, keeping and checking that the slice is being executed on
the coordinator by the conditions flotype = FLOW_SINGLETON and segindex < 0.
We get rid of the need to fill in the flow for a shared scan in ORCA,
and now we can remove the function that did this.

RekGRpth committed:
Merge branch 'adb-6.x-dev' into ADBDEV-3333-3

InnerLife0 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -2405,21 +2425,29 @@ shareinput_mutator_xslice_1(Node *node, PlannerInfo *root, bool fPop)
 
-		Assert(sisc->scan.plan.flow);
-		if (sisc->scan.plan.flow->flotype == FLOW_SINGLETON)
+		if (qds == 1)
```
Is it possible to simplify this?
```
Flow *flow = sisc->scan.plan.flow;
if (qds == 1 || (qds == -1 && flow && flow->flotype == FLOW_SINGLETON && flow->segindex < 0))
    ctxt->qdShares = list_append_unique_int(ctxt->qdShares, sisc->share_id);
```
More, I think we need an in-code comment about what's going on here.

RekGRpth replied:
done

InnerLife0 replied:
The value of such comments is almost zero. Do you think it's correctly described to somebody who will look at code in future?

RekGRpth replied:
removed comments

InnerLife0 replied:
removed or replaced?

RekGRpth replied:
comment about logic was before function https://github.com/arenadata/gpdb/pull/498/files#diff-a4950bc4a9e9552824f7ba4f6bf300b64335fb668074e51f7b9ee3ac30fe7a78L2388

InnerLife0 replied:
Does this line explain why we have such condition here?

RekGRpth replied:
this is only saving previous logic

InnerLife0 replied:
We're going in circles.
We had a voice talk about why it's bad that there no comment here and why we need it.
Instead of placing a comment here, you prefer to find 1000 and 1 reason to not do this. Why you afraid to do this?

RekGRpth committed:
simplify and comment

Stolb27 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -2236,6 +2250,12 @@ shareinput_peekmot(ApplyShareInputContext *ctxt)
 }
 
+static int
+shareinput_peekqds(ApplyShareInputContext *ctxt)
```
In my opinion, meaning of qds term and its possible values is very unobvious. Maybe use enum?

RekGRpth replied:
> In my opinion, meaning of qds term and its possible values is very unobvious.

qds stands for query dispatcher slice

RekGRpth replied:
> Maybe use enum?

done

Stolb27 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -2559,8 +2586,6 @@ shareinput_mutator_xslice_3(Node *node, PlannerInfo *root, bool fPop)
 
 		if (list_member_int(ctxt->qdShares, sisc->share_id))
 		{
-			Assert(sisc->scan.plan.flow);
-			Assert(sisc->scan.plan.flow->flotype == FLOW_SINGLETON);
```
It seems that we are still able to validate that slice under motId is singleton.

RekGRpth replied:
only if `sisc->scan.plan.flow != NULL`

RekGRpth replied:
done

Stolb27 commented on src/test/regress/sql/with_clause.sql:
```diff
@@ -390,3 +390,18 @@ with recursive rcte(x,y) as
 )
 select * from rcte limit 10;
+
+-- ensure orca does not fail when one cte on coordinator has correct flow and another - not
```
I think it would be nice to point, that problem may be reproduced on build with assert only. As far as this test checks only orca behavior, maybe move it to gporca test file?

RekGRpth replied:
> I think it would be nice to point, that problem may be reproduced on build with assert only.

done

RekGRpth replied:
> As far as this test checks only orca behavior, maybe move it to gporca test file?

How? `gporca.sql` has `gporca.out` and `gporca_optimizer.out`

RekGRpth committed:
comment

RekGRpth committed:
enum and comments and tests coments

InnerLife0 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -2219,15 +2219,34 @@ collect_shareinput_producers(PlannerInfo *root, Plan *plan)
+	Assert(motion->plan.lefttree);
+	Assert(motion->plan.lefttree->flow);
+
+	Flow *flow = motion->plan.lefttree->flow;
```
This line should be moved upper (above asserts).

RekGRpth replied:
> This line should be moved upper (above asserts).

it may cause segfault

InnerLife0 replied:
Then we should rewrite this.

RekGRpth replied:
done

InnerLife0 replied:
OK.
I think we need to check the whole new code with pgindent after finish.

RekGRpth replied:
> we need to check the whole new code with pgindent

done

RekGRpth committed:
fix

RekGRpth committed:
indent

RekGRpth committed:
rewrite

RekGRpth committed:
assert

RekGRpth committed:
indent

RekGRpth committed:
indent

RekGRpth committed:
fix

RekGRpth committed:
rename

RekGRpth committed:
fix

RekGRpth committed:
optimize

RekGRpth committed:
optimize

RekGRpth committed:
assert

RekGRpth committed:
optimize

RekGRpth committed:
indent

RekGRpth committed:
comments

RekGRpth committed:
indent

InnerLife0 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -2624,7 +2733,7 @@ apply_shareinput_xslice(Plan *plan, PlannerInfo *root)
 	ctxt->sliceMarks = palloc0(ctxt->producer_count * sizeof(int));
 
-	shareinput_pushmot(ctxt, 0);
+	shareinput_pushmot(ctxt, &fakeMotion);
```
I understand why it's safe to pass a ptr to local variable here, but it still looks dirty. The reason - current code can't guarantee that motStack is empty at the end.

RekGRpth replied:
Was that https://github.com/arenadata/gpdb/pull/498/commits/069f0c3a4ac8dba5c093f62d9f413e5cc209d5a3#diff-a4950bc4a9e9552824f7ba4f6bf300b64335fb668074e51f7b9ee3ac30fe7a78L2626-L2630 better?

RekGRpth replied:
Also, I may suggest following
```c
List *
lcons_obj(void **datum, Size size, List *list)
{
    *datum = palloc(size);
    return lcons(*datum, list);
}
List *
list_delete_first_obj(List *list)
{
    void *datum = linitial(list);
    pfree(datum);
    return list_delete_first(list);
}
struct Some
{
    int motionID;
    FlowType flotype;
    int segindex;
}
static void
shareinput_pushmot(ApplyShareInputContext *ctxt, Motion *motion)
{
    struct Some *some;
    ctxt->motStack = lcons_obj(&some, sizeof(*some), ctxt->motStack);
    some->motionID = motion->motionID;
    some->flowtype = motion->plan.lefttree->flow->flowtype;
    some->segindex = motion->plan.lefttree->flow->segindex;
}
static void
shareinput_popmot(ApplyShareInputContext *ctxt)
{
    list_delete_first_obj(ctxt->motStack);
}
static struct Some *
shareinput_peeksome(ApplyShareInputContext *ctxt)
{
    return linitial(ctxt->motStack);
}
static int
shareinput_peekmot(ApplyShareInputContext *ctxt)
{
    return shareinput_peeksome(ctxt->motStack)->motionID;
}
struct Some *some = shareinput_peeksome(ctxt);
if (some->flotype == FLOW_SINGLETON && some->segindex < 0)
{
    ...
}
```

InnerLife0 replied:
> Was that [069f0c3#diff-a4950bc4a9e9552824f7ba4f6bf300b64335fb668074e51f7b9ee3ac30fe7a78L2626-L2630](https://github.com/arenadata/gpdb/commit/069f0c3a4ac8dba5c093f62d9f413e5cc209d5a3#diff-a4950bc4a9e9552824f7ba4f6bf300b64335fb668074e51f7b9ee3ac30fe7a78L2626-L2630) better?

This was better.

RekGRpth replied:
> This was better.

done

InnerLife0 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -2559,8 +2568,8 @@ shareinput_mutator_xslice_3(Node *node, PlannerInfo *root, bool fPop)
-			Assert(sisc->scan.plan.flow);
-			Assert(sisc->scan.plan.flow->flotype == FLOW_SINGLETON);
+			Assert(shareinput_peekflow(ctxt));
+			Assert(shareinput_peekflow(ctxt)->flotype == FLOW_SINGLETON);
```
I should leave a comment here, but I can't come up with better solution.

RekGRpth committed:
rewrite

InnerLife0 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -2610,12 +2619,112 @@ shareinput_mutator_xslice_4(Node *node, PlannerInfo *root, bool fPop)
+{
+	Query	   *query = root->parse;
+
+	if (query->parentStmtType)
```
1) Please add a direct comparison with PARENTSTMTTYPE_NONE.
2) What constant covers SELECT INTO cases?

RekGRpth replied:
> Please add a direct comparison with PARENTSTMTTYPE_NONE.

done

RekGRpth replied:
> What constant covers SELECT INTO cases?

https://github.com/arenadata/gpdb/blob/e69509affc7d96b7298b9ae80582c974cd4b0d4a/src/backend/utils/cache/plancache.c#L701

InnerLife0 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -2610,12 +2619,112 @@ shareinput_mutator_xslice_4(Node *node, PlannerInfo *root, bool fPop)
+	{
+		fakeFlow.flotype = FLOW_SINGLETON;
+		fakeFlow.segindex = -1;
+	}
```
Where is the else condition? Do you rely on fact that first enum value is zero?

RekGRpth replied:
> Where is the else condition? Do you rely on fact that first enum value is zero?

done

RekGRpth committed:
indent

RekGRpth committed:
direct cmp

RekGRpth committed:
else

InnerLife0 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -2610,12 +2619,115 @@ shareinput_mutator_xslice_4(Node *node, PlannerInfo *root, bool fPop)
+			ListCell   *lc = list_head(mt->resultRelations);
+			Oid			reloid = getrelid(lfirst_int(lc), query->rtable);
+
+			if (GpPolicyFetch(reloid)->ptype != POLICYTYPE_ENTRY)
```
It seems we can heavily simplify this conditions.
```suggestion
			return (GpPolicyFetch(reloid)->ptype == POLICYTYPE_ENTRY);
```

RekGRpth replied:
> It seems we can heavily simplify this conditions.

I removed the else, so now it can't be so simplified.

InnerLife0 replied:
Sometimes I think you joking me.

What was the reason of removal?

RekGRpth replied:
> What was the reason of removal?

I revised the conditions in the executor.

InnerLife0 replied:
Please give a link or explain it somehow.

RekGRpth replied:
> Please give a link or explain it somehow.

I tried to describe it in the header:

> 1) For `CTAS`, `SELECT INTO`, `COPY INTO`, `REFRESH MATERIALIZED VIEW`, [the root `slice` is not executed on the `coordinator`](https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/executor/execMain.c#L5231-L5247).
> 2) For the modification node, the table distribution policy is checked, and if the table is not master-only, then [the root `slice` is not executed on the `coordinator`](https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/executor/execMain.c#L5051-L5116).
> 3) Parallel cursors can only be generated by the postgres planner, which means [it will set the correct `flow`](https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/executor/execMain.c#L1809-L1843).
If it is not a `singleton`, then the root `slice` is not executed on the `coordinator`.
If this is a `singleton`, but the `locus` is not `Entry`, not `General`, and not `SingleQE`, then the root `slice` is also not executed on the `coordinator`.
> 4) The postgres planner fills `flow` correctly.
If it is not a `singleton`, then the root `slice` is not executed on the `coordinator`.
If this is a `singleton` on a segment, then the root `slice` is also not executed on the `coordinator`.
> 5) In all other cases, the root `slice` is executed on the `coordinator`.

InnerLife0 replied:
I don't see any explanation here about why this conditions (or their subset) can't use `else if`.

RekGRpth replied:
> I don't see any explanation here about why this conditions (or their subset) can't use `else if`.

By default, the root slice is executed on the coordinator, and only under certain conditions is it not.

InnerLife0 replied:
> By default, the root slice is executed on the coordinator, and only under certain conditions is it not.

How this explains my question about `else` operator?

RekGRpth replied:
[Here](https://github.com/arenadata/gpdb/pull/498/files#diff-a4950bc4a9e9552824f7ba4f6bf300b64335fb668074e51f7b9ee3ac30fe7a78R2633-R2636) the else operator is not needed

InnerLife0 replied:
Can you clarify why it was necessary to remove else condition in this line?

RekGRpth replied:
Because if `if` ends with a `return`, then the `else` is redundant.

InnerLife0 replied:
True.
What about other cases? Can you explain them in their threads?
UPD. I mean, initially I didn't ask about else in this block of code. Why you removed others (which made my suggestions irrelevant)?

RekGRpth replied:
[Here](https://github.com/arenadata/gpdb/pull/498/files#diff-a4950bc4a9e9552824f7ba4f6bf300b64335fb668074e51f7b9ee3ac30fe7a78R2655-R2657) we can use `else`.

RekGRpth replied:
> What about other cases?

I'm thinking about it...

InnerLife0 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -2610,12 +2619,115 @@ shareinput_mutator_xslice_4(Node *node, PlannerInfo *root, bool fPop)
+		DML		   *dml = (DML *) plan;
+		Oid			reloid = getrelid(dml->scanrelid, query->rtable);
+
+		if (GpPolicyFetch(reloid)->ptype != POLICYTYPE_ENTRY)
```
return (GpPolicyFetch(reloid)->ptype == POLICYTYPE_ENTRY);

InnerLife0 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -2610,12 +2619,115 @@ shareinput_mutator_xslice_4(Node *node, PlannerInfo *root, bool fPop)
+		DML		   *dml = (DML *) plan;
+		Oid			reloid = getrelid(dml->scanrelid, query->rtable);
+
+		if (GpPolicyFetch(reloid)->ptype != POLICYTYPE_ENTRY)
```
```suggestion
		return (GpPolicyFetch(reloid)->ptype == POLICYTYPE_ENTRY);
```

RekGRpth replied:
I removed the else, so now it can't be so simplified.

RekGRpth replied:
[Here](https://github.com/arenadata/gpdb/pull/498/files#diff-a4950bc4a9e9552824f7ba4f6bf300b64335fb668074e51f7b9ee3ac30fe7a78R2655-R2657) I removed the `else`, because the executor always checks for parallel cursors at the end.

RekGRpth replied:
Although it is in this place that `else` can be left, because both conditions are incompatible.

RekGRpth replied:
But even in this case, it is impossible to simplify as above.

InnerLife0 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -2610,12 +2619,115 @@ shareinput_mutator_xslice_4(Node *node, PlannerInfo *root, bool fPop)
+
+		if (root->glob->is_parallel_cursor)
+		{
+			if (flow->flotype == FLOW_SINGLETON &&
```
```suggestion
			return (flow->flotype == FLOW_SINGLETON &&
```

RekGRpth replied:
I removed the else, so now it can't be so simplified.

InnerLife0 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -2610,12 +2619,115 @@ shareinput_mutator_xslice_4(Node *node, PlannerInfo *root, bool fPop)
+			}
+		}
+
+		if (flow->flotype != FLOW_SINGLETON || flow->segindex >= 0)
```
```suggestion
		return (flow->flotype == FLOW_SINGLETON || flow->segindex < 0);
```

RekGRpth replied:
This can be done, but style and comment will be lost.

InnerLife0 replied:
1) How can we _lose_ the style? I mean, in general?
2) Do we write a comment for a code or a code for a comment?

RekGRpth replied:
> How can we _lose_ the style? I mean, in general?

I return false under some conditions, and the comment describes it in more detail.

RekGRpth replied:
> return (flow->flotype == FLOW_SINGLETON || flow->segindex < 0);

It's not the same condition!

RekGRpth replied:
> return (flow->flotype == FLOW_SINGLETON || flow->segindex < 0);

It's not the same condition!

Maybe you mean
```c
return (flow->flotype == FLOW_SINGLETON && flow->segindex < 0);
```
?

InnerLife0 replied:
> It's not the same condition!

Please provide a correct one then.


_Added after update of quoted message:_
> Maybe you mean

Yes, it may work.

InnerLife0 replied:
> 2. Do we write a comment for a code or a code for a comment?

This question is still relevant.

RekGRpth replied:
> Please provide a correct one then.

See previous comment above.

RekGRpth replied:
> > 2. Do we write a comment for a code or a code for a comment?
> 
> This question is still relevant.

A comment for a code, of course.

InnerLife0 replied:
> A comment for a code, of course.

Then we shouldn't use unavailability of changing or moving comment as a reason. Thank you.

InnerLife0 commented:
> To determine where the slice will be executed, we can use the node's flow under Motion, keeping and checking that the slice is being executed on the coordinator by the conditions flotype = FLOW_SINGLETON and segindex < 0.

Is this sentence still relevant or we always use Motion's flow now?

> 5. In all other cases, the root slice is executed on the coordinator.
We get rid of the need to fill in the flow for a shared scan in ORCA, and now we can remove the function that did this.

Misformatting.

RekGRpth committed:
remove else

RekGRpth committed:
remove else

RekGRpth committed:
indent

RekGRpth commented:
> > 9\. In all other cases, the root slice is executed on the coordinator.
> >     We get rid of the need to fill in the flow for a shared scan in ORCA, and now we can remove the function that did this.
> 
> Misformatting.

fixed

RekGRpth commented:
> Is this sentence still relevant or we always use Motion's flow now?

Yes we always use flow under Motion now.

RekGRpth commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -2610,12 +2619,114 @@ shareinput_mutator_xslice_4(Node *node, PlannerInfo *root, bool fPop)
+		}
+	}
+
+	if (plan->flow)
```
[Here](https://github.com/arenadata/gpdb/pull/498/files#diff-a4950bc4a9e9552824f7ba4f6bf300b64335fb668074e51f7b9ee3ac30fe7a78R2671-R2673) I removed the `else`, because the executor always checks for parallel cursors at the end.

RekGRpth committed:
rewrite conditions

RekGRpth committed:
indent

InnerLife0 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -2666,44 +2670,45 @@ root_slice_is_executed_on_coordinator(Plan *plan, PlannerInfo *root)
 				 * interact with the retrieve connections.
 				 */
-				return false;
+				result = true;
```
So here we can overwrite `result` by true if it was set to false somewhere upper. Is it correct?
What was the reason of this commit? Why it's important to use variable instead of return?

RekGRpth replied:
> So here we can overwrite `result` by true if it was set to false somewhere upper. Is it correct?

Yes, as I understand it, the executor does just that. https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/executor/execMain.c#L1809-L1843

RekGRpth replied:
> What was the reason of this commit? Why it's important to use variable instead of return?

To allow the result to be overwritten.

InnerLife0 replied:
> Yes, as I understand it, the executor does just that.

It's a good answer with the link. I don't need to break my mind searching for a reason.
Still better to comment this in time and not on demand, but anyway, thank you. I'll take a look.

RekGRpth replied:
The funny thing is that I have not yet found a test to confirm the need for all [this](https://github.com/arenadata/gpdb/blob/de054aa30295842860fb8c405db03e8d01711c3a/src/backend/cdb/cdbmutate.c#L2616). When I replaced assignment with assert, none of the tests broke.

RekGRpth replied:
And in general, I believe that the executor, not the planner, should check the plan for validity, because if the planner understands that the plan is invalid, then why shouldn't it rebuild the plan?

InnerLife0 reviewed (approved):
Let it be
